### PR TITLE
Automatically enable worker ENVIRONMENT when needed

### LINKED ISSUE #14162: Auto-include required code when using `-pthread -s ENVIRONMENT=web`
Tracking issue for @kripken's idea in #14151:

> Allow just setting ENVIRONMENT=node in combination with pthreads, as
that is unambiguous, and should not be broken.

> Perhaps we should do the same with pthreads outside of node, however
because of the node support code it is trivial to do it in node. I suspect it
would be more work otherwise.

Instead of showing an error telling used to use `-s ENVIRONMENT=web,worker` instead of `-s ENVIRONMENT=web`, we can just do the right thing and include the required code when `-pthread` is provided anyway.

### PR DESCRIPTION
**Pull request by RReverser**
- Similar to #14151 but for more combos.
 - Fixes #14162.

### CODE REVIEW

**Comment by sbc100** [comment]
Is seems a little of to me to automatically add environment support when the user explicitly sets the supported environment. 

Its somewhat contradictory for example if the user only wants the `shell` environment to then inject `worker` too.   

Maybe its good for convenience but perhaps it at least warrents a warning.

e.g. "You explictly specified `-sENVIRONMENT` but you didn't include `worker`.   Automatically adding `worker` to this list due to use of `-pthread` and/or `-sWASM_WORKERS`".?

**Comment by RReverser** [comment]
Idk, in terms of DX I'm personally a big fan of automatically doing the right thing vs erroring out (or even warning).

If user already specified `-pthread -s ENVIRONMENT=web`, what do we gain from them having to also spell out `...,worker`? They already opted in to `-pthread`, so it seems pretty obvious that by "Web" they mean "Web and whatever implementation details you need for the pthread API".

We don't do this for other feature-specific implementation details, eg we don't have granularity for `-s ENVIRONMENT=audioworklet` (which is different from workers, but we simply detect it automatically), so why is `worker` special?

Plus, there is already a precedent with Node.js already automatically handling workers linked above.

After this PR the only reason to ever explicitly set `-s ENVIRONMENT=worker` is if the user wants to run Emscripten's main thread inside a Web Worker, but I don't think we should require it or warn on it when it's only necessary as an implementation detail for Emscripten's own workers.

**Comment by RReverser** [comment]
> After this PR the only reason to ever explicitly set `-s ENVIRONMENT=worker` is if the user wants to run Emscripten's main thread inside a Web Worker, but I don't think we should require it or warn on it when it's only necessary as an implementation detail for Emscripten's own workers.

Although, come to think of it, we could auto-detect that case in exactly the same way Audio Worklets are, and then we could deprecate separate `-s ENVIRONMENT=worker` altogether.

**Comment by sbc100** [comment]
> -s ENVIRONMENT=worker

Yes, I'd be up for having `-sENVIRONMENT=worker` *just* mean that you want to run the main runtime in a worker.

What do you mean by `auto-detect that case in exactly the same way Audio Worklets are`.   I assume there is some extra cost/complexity to supporting running the main runtime in worker?

In any case this change LGTM, but can you also remove `worker` from the default `-sENVIRONMENT` in that case?   (and from any multi-pthreading test cases that specify `worker` explicitly).  i.e. lets from all the explicit usage if we are making it implicit.

**Comment by RReverser** [comment]
> What do you mean by `auto-detect that case in exactly the same way Audio Worklets are`. I assume there is some extra cost/complexity to supporting running the main runtime in worker?

Ah, that's a good question, might need to check how much of that complexity is for main runtime in worker vs for supporting shared memory. Either way, that would be outside of this PR.

> In any case this change LGTM, but can you also remove `worker` from the default `-sENVIRONMENT` in that case?

Do you mean from https://github.com/emscripten-core/emscripten/blob/8527d8dc8459f4350c62248a6eeaf1cffcd7dc9e/src/settings.js#L667? But then it would be a breaking change for users who do run their Emscripten-generated code in a Web Worker?

> (and from any multi-pthreading test cases that specify `worker` explicitly)

This one makes sense, yeah.

**Comment by RReverser** [comment]
> Ah, that's a good question, might need to check how much of that complexity is for main runtime in worker vs for supporting shared memory. Either way, that would be outside of this PR.

Looks like there are only 3 (?) usages of `ENVIRONMENT_IS_WORKER` that aren't guarded by features like pthreads, so while it's not zero, the extra cost is pretty low. Either way, we can decide that later.

**Comment by RReverser** [comment]
> Can you remove `worker` from the `var ENVIRONMENT = 'web,webview,worker,node';` line in settings.js ?

As I mentioned above, unlike rest of this PR so far, that would be a breaking change. I'd rather keep it separate if we decide to do that as well.

File: ChangeLog.md [hunk 1]
```diff
@@ -20,10 +20,12 @@ See docs/process.md for more on how version tagging works.
 
 4.0.11 (in development)
 ----------------------
-
+- The `ENVIRONMENT` setting will now be automatically updated to include worker
```
Comment (sbc100) [inline ChangeLog.md]: Backticks around "worker"?

**Comment by sbc100** [comment]
> > Can you remove `worker` from the `var ENVIRONMENT = 'web,webview,worker,node';` line in settings.js ?
> 
> As I mentioned above, unlike rest of this PR so far, that would be a breaking change. I'd rather keep it separate if we decide to do that as well.

sgtm

File: ChangeLog.md (hunk 1 above)
Comment (RReverser) [reply ChangeLog.md]: I guess, yeah.

**Comment by RReverser** [comment]
> and from any multi-pthreading test cases that specify `worker` explicitly

That was a good call, one of them actually caught an edge case I didn't consider.
